woocommerce/qit-cli
# Improve QIT self-tests

This PR adds the groundwork for an extensible and simple way of creating and running QIT self-tests.

This PR adds a Runner for the self-tests. When you run `php QitSelfTests.php run`, it will:

- For each test type directory (eg: `activation`)
  - For each test (eg: `activation/test-1`)
    - Create the zip of the SUT
    - Create a PHPUnit test file
    - Dispatch a QIT test with `--wait` and `--json` parameters
    - Compare the JSON result with a previously saved snapshot

If the JSON result is different from the snapshot, the test will fail. All tests are dispatched in parallel, and the results are processed as they come in. When a result is available, it will compare the JSON result with a previously saved snapshot.

A more comprehensive documentation can be found [here](https://github.com/woocommerce/qit-cli/blob/76aee2cbd053d07a35547d61851d3730f976d8d3/_tests/README.md).

### Testing instructions
- Checkout this branch
- You might want to use the `staging` environment in QIT CLI
- Open the `_tests` directory in your terminal
- Run `php QITSelfTests.php` and assert that it runs the tests as expected
- Change the SUTs, either by creating a new error in some of the activation tests, or a new security violation to security tests
- Run the tests again and assert that it fails
- Re-generate the snapshots with `php QITSelfTests.php update`
- Assert that you can commit a new snapshot with the expected change
